### DISCUSSION TITLE
Duplicate Command Opcodes and Thread attr

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#1287] [Q&A] [answered]
Hello, 

I notice that some of my commands have the same opcode. See below the ASSERT messages. I played with following config arguments but wasn't able to solve it: 
CmdDispatcherComponentCommandPorts in AcConstants.ini (These are the num of ports so it shouldn't solve it)
CMD_DISPATCHER_DISPATCH_TABLE_SIZE in CommandDispatcherImplCfg.hpp (I don't understand this exactly, this is the table which stores commands to be dispatched once it has received it and not the size that it will use in genrating xml file)

I can also see #define FwOpcodeType U32   which means it should be able to store a lot of commands. Can you help me find the solution? 

```
Assert file "/home/sheershak/Desktop/F-Prime/fprime/Svc/CmdDispatcher/CommandDispatcherImpl.cpp": Line: 51 421
EVENT: (2001) (2:1645290176,899012) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A5 registered to port 10 slot 72
Assert file "/home/sheershak/Desktop/F-Prime/fprime/Svc/CmdDispatcher/CommandDispatcherImpl.cpp": Line: 51 422
EVENT: (2001) (2:1645290176,899089) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A6 registered to port 10 slot 73
Assert file "/home/sheershak/Desktop/F-Prime/fprime/Svc/CmdDispatcher/CommandDispatcherImpl.cpp": Line: 51 423
EVENT: (2001) (2:1645290176,899164) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A7 registered to port 10 slot 74
Assert file "/home/sheershak/Desktop/F-Prime/fprime/Svc/CmdDispatcher/CommandDispatcherImpl.cpp": Line: 51 424
EVENT: (2001) (2:1645290176,899238) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A8 registered to port 10 slot 75
Assert file "/home/sheershak/Desktop/F-Prime/fprime/Svc/CmdDispatcher/CommandDispatcherImpl.cpp": Line: 51 425
EVENT: (2001) (2:1645290176,899312) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A9 registered to port 10 slot 76
Assert file "/home/sheershak/Desktop/F-Prime/fprime/Svc/CmdDispatcher/CommandDispatcherImpl.cpp": Line: 51 426
EVENT: (2001) (2:1645290176,899386) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01AA registered to port 10 slot 77

```

Thanks

**Comment by Joshua-Anderson** [comment, marked as answer]
This command dispatch table maps command opcode to component ports that should be called to dispatch the command.

The most common cause for warnings like what you're seeing is if a component's base id in your component topology overlaps with other components. Since it looks like the assert is about command opcodes 421-424, I'd take a look in the dictionary and see what's getting allocated those opcodes. If there is an overlap, adjust your base id and base id window to separate the opcode ranges of the two components.

**Reply by sheershak-agarwal** [reply]
so is base_id_window = max (number of command,number of events,number of channels)?

**Reply by Joshua-Anderson** [reply]
Correct. Command opcodes, event opcodes, and telemetry opcodes within a component each have their own individual opcode ranges, starting with the supplied `base_id` number. So `base_id_window` must be sized to hold the largest of those three ranges.

**Reply by sheershak-agarwal** [reply]
thank you

**Comment by sheershak-agarwal** [comment]
One other question, in Topology.cpp when we use component init and start function. the init function has quesueDepth and instance. What does it mean? 
also start function has NATIVE_UINT_TYPE priority, NATIVE_UINT_TYPE stackSize, NATIVE_UINT_TYPE cpuAffinity, NATIVE_UINT_TYPE identifier. can you explain a little about them too so I know what exactly they mean(what can be the max values and whats the best way to find a stack size for it)? or guide me to a document.
Thanks

**Reply by Joshua-Anderson** [reply]
There is [C++ Documentation](https://nasa.github.io/fprime/v2.1.0/UsersGuide/api/c++/html/index.html), but you're right that there currently isn't any documentation on those functions. For init, `queueDepth` is used to define the maximum size of the component's message queue. 

For start:

- priority: task priority, OS specific meaning
- stack size: task stack size
- cpuAffinity: pin task to specific CPU (defaults to no pinning)
- identifier: numeric identifier for task (optional, for use by user)

All of these can be safely be set to Os::TASK_DEFAULT if you want to leave these to the default behavior.

**Comment by sheershak-agarwal** [comment]
Hi @Joshua-Anderson,

I am getting these assert errors at the Svc::Sched port in several ports (mostly running at 1Hz or higher).  The line 9208 and 1283 are in schedIn_handlerBase function and 8 means that the QUEUE is full.
```
Assert file "/home/sheershak/F-Prime/App/build-fprime-automatic-raspberrypi/App/MissionStateManager/MissionStateManagerComponentAc.cpp": Line: 1283 8
Assert file "/home/sheershak/F-Prime/App/build-fprime-automatic-raspberrypi/App/Gnc/GncComponentAc.cpp": Line: 9208 8
Assert file "/home/sheershak/F-Prime/App/build-fprime-automatic-raspberrypi/App/Gnc/GncComponentAc.cpp": Line: 9208 8
```

To overcome this error I tried to increase the queueDepth in init call and it did help but only for a short amount of time.  Then I played around with the priority parameter in the start call but that didn't help either. (and I don't think cpuafinity will help either).

How can I solve these assert errors? Btw, my sched ports are initialized with kind="async_input". Will changing to sync_input help?

**Reply by sheershak-agarwal** [reply]
I made the input as sync input and started see these warning. Now I'm thinking that the error is happening in RateGroups.
EVENT: (162) (2:1647375395,270833) WARNING_HI: (RG4) RateGroupCycleSlip : Rate group cycle slipped on cycle 0
EVENT: (162) (2:1647375395,844464) WARNING_HI: (RG4) RateGroupCycleSlip : Rate group cycle slipped on cycle 1
EVENT: (162) (2:1647375396,419328) WARNING_HI: (RG4) RateGroupCycleSlip : Rate group cycle slipped on cycle 2
EVENT: (162) (2:1647375396,993395) WARNING_HI: (RG4) RateGroupCycleSlip : Rate group cycle slipped on cycle 3
EVENT: (162) (2:1647375397,567950) WARNING_HI: (RG4) RateGroupCycleSlip : Rate group cycle slipped on cycle 4

**Reply by sheershak-agarwal** [reply]
and later I get EVENT: (21) (2:1647376368,891076) WARNING_HI: (health) HLTH_PING_WARN : Ping entry RG4 late warning
and a FATAL event.

**Reply by LeStarch** [reply]
It seems that the endpoint isn't running.  Have you started the component that is being triggered by the rategroup?  If so, is the call returning?  It could also be cause by an infinite loop on the sched handler.

**Reply by sheershak-agarwal** [reply]
Yes, I have started it like this (with priorities)
msm.start(97);
gnc.start(98);

The call checks something, doesn't return anything. and not running an infinite loop.

**Reply by sheershak-agarwal** [reply]
These are returned on the Channels: (I see two rg4 slips on every 4 BD_Cycles), my base clock is 4Hz


Last Sample Time | Channel Id | Channel Name | Channel Value
-- | -- | -- | --
2022-03-15T20:47:03.832Z | 0x15 | health.PingLateWarnings | 1
2022-03-15T20:45:27.308Z | 0x65 | rateGroup1Comp.RgMaxTime | 527902 us
2022-03-15T20:45:23.778Z | 0x79 | rateGroup2Comp.RgMaxTime | 346 us
2022-03-15T20:46:03.801Z | 0x8d | rateGroup3Comp.RgMaxTime | 907 us
2022-03-15T20:47:36.865Z | 0xa1 | rateGroup4Comp.RgMaxTime | 5769938 us
2022-03-15T20:47:44.736Z | 0xa2 | rateGroup4Comp.RgCycleSlips | 269
2022-03-15T20:47:44.853Z | 0xb5 | blockDrv.BD_Cycles | 564

**Reply by sheershak-agarwal** [reply]
And I notice the RgMaxTime for rg1 and rg4 are really high.

**Reply by sheershak-agarwal** [reply]
I found the issue for two threads, which is that they are sleeping and the scheduler keeps filling the queue. So my only question now is what happens to the new element coming to the queue if the queue is full(does it discard it or keeps it and removes the oldest element). and when the thread is awake again, does it perform all the elements or discards it?

**Reply by LeStarch** [reply]
The basic overflow behavior for the queue is to assert the system.  That is the assert you see above (`...Ac.cpp: <line> 8`).  You can tell a queue to drop by setting the `full` attribute of the port in the component model.  In the XML model format it is:

```
<port .... full="drop">
```
full may also be set to:

1. assert: default, you'll get an assert with argument 8 as shown above on full
2. drop: drops messages on full
3. block: blocks when full.  This will likely cause something else to overflow (or block).  Use with caution.

**Reply by sheershak-agarwal** [reply]
Really helpful. Thank you

**Reply by LeStarch** [reply]
I am really glad that helped!  Thanks for the feedback.
